### PR TITLE
Cli

### PR DESCRIPTION
**Pull request by Jannes123**
Usable cli for inasafe that produces shapefiles and pdf report if impact layer.

### CODE REVIEW

**Comment by Jannes123** [comment]
Please review @timlinux

**Comment by Jannes123** [comment]
Please review @timlinux

**Comment by ismailsunni** [comment]
Hi @Jannes123 
can you give more information how to do install it?
I follow the steps in `usage.txt`:
```# INSTALL on Ubuntu with:
# chmod ug+x inasafe
# sudo ln -s `pwd`/inasafe  /usr/bin```

But, when I wanted to run it, I got:
```
dogerice: inasafe 
Traceback (most recent call last):
  File "/usr/bin/inasafe", line 21, in <module>
    from safe.impact_functions.registry import Registry
ImportError: No module named safe.impact_functions.registry
```

I think it's missing a path to inasafe. May be you can add how to do it in the usage.txt

File: bin/inasafe [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+    :rtype: list
+    """
+    LOGGER.debug('get IF list')
+    registry = Registry()
+    LOGGER.debug(registry.impact_functions)
+    return registry.impact_functions
+
+
+def show_names(impact_functions):
+    """Prints a list of strings.
```
Comment (timlinux) [inline bin/inasafe]: print a list of impact function names?

File: bin/inasafe [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+     the Free Software Foundation; either version 2 of the License, or
+     (at your option) any later version.
+
+    Note :
+        cli: command line interface
+"""
+
+__author__ = 'Jannes Engelbrecht'
+__date__ = '16/04/15'
+from docopt import docopt, DocoptExit
```
Comment (timlinux) [inline bin/inasafe]: Please pay attention to import order:

* builtins
* third party
* project

See https://github.com/AIFDR/inasafe/wiki/Coding-Standards#ordering-of-imports

File: bin/inasafe [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+usage_dir = os.environ['InaSAFEQGIS'] + '/bin'
+usage = r""
+usage_file = file(os.path.join(usage_dir, 'usage.txt'))
+for delta in usage_file:
+    usage += delta
+LOGGER = logging.getLogger('InaSAFE')
+QGIS_APP, CANVAS, IFACE, PARENT = get_qgis_app()
+
+
+class CliArgs(object):
```
Comment (timlinux) [inline bin/inasafe]: Is this naming convention required by docopts? If not please call CommandLineArguments or similar (to follow our coding conventions).

File: bin/inasafe [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+        self.impact_function = _arguments_['--impact-function']
+        self.report_template = _arguments_['--report-template']
+        # optional arguments
+        if not _arguments_['--extent'] is None:
+            self.extent = _arguments_['--extent'].replace(',', '.').split(':')
+        else:
+            LOGGER.debug('no extent specified')
+
+
+def get_ifunction_list():
```
Comment (timlinux) [inline bin/inasafe]: No abbreviations please

File: bin/inasafe [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+        An impact layer object is created and used to write a shapefile.
+        The shapefile path is given by user and used by build_report
+        function to read from.
+
+    :param cli_arguments: User inputs.
+    :type cli_arguments: CliArgs
+
+    :raises: Exception
+    """
+    qhazard = get_hazard(cli_arguments)
```
Comment (timlinux) [inline bin/inasafe]: What is the q suffix for? Can we remove it?

**Comment by Jannes123** [comment]
@ismailsunni the script assumes same environment as inasafe test suite.  So you have to source run-env-linux.sh and check that you can import safe from python.  I will add this to usage.

**Comment by Jannes123** [comment]
@timlinux there is an issue with the keywords when I do a download before running an IF.  Inasafe creates an .xml file from .keywords but the Analysis throws the following error:
inasafe  --hazard=/home/jannes/gitwork/inasafe/safe/test/data/hazard/continuous_flood_20_20.asc --download --layers=roads --extent=106,8055918:-6,1973263:106,8436206:-6,1586401 --impact-function=FloodRasterRoadsFunction --report-template=/home/jannes/gitwork/inasafe/resources/qgis-composer-templates/inasafe-portrait-a4.qpt  --output-file=/home/jannes/experiment/flood_on_buildings.shp
QGraphicsScene::addItem: item has already been added to this scene
insafe

roads
temp directory: /tmp/tmp1Z8GjC/exposure
Exposure download extent is valid
download successful
Loading a style file that was saved with an older version of qgis (saved in 1.9.0-Master, loaded in 2.8.2-Wien). Problems may occur.
hazard layer is VALID
Loading a style file that was saved with an older version of qgis (saved in 2.3.0-Master, loaded in 2.8.2-Wien). Problems may occur.
exposure layer is VALID
Warning 1: Field TYPE of width 255 truncated to 254.
2015-06-10 17:25:52,304 - InaSAFE - ERROR - An exception occurred when running the impact analysis.
Traceback (most recent call last):
  File "/home/jannes/.qgis2/python/plugins/inasafe/safe/utilities/analysis.py", line 965, in run_analysis
    self.impact_layer = calculate_impact(self.impact_function)
  File "/home/jannes/.qgis2/python/plugins/inasafe/safe/engine/core.py", line 222, in calculate_impact
    exp = result_layer.keywords['exposure'].title()
KeyError: 'exposure'
'NoneType' object has no attribute 'write_to_file'

Unspecified run-time error.

**Comment by timlinux** [comment]
@jannes we need @Gustry to update the keywords file before that will work We have the issue that the new and old keywords are incompatible so we cant deploy them to the server until they are. Please test against a dataset where you have prepared the keywords against a dataset manually using develop's eyword wizard. Thanks.

**Comment by timlinux** [comment]
Sorry I closed your PR by accident!

**Comment by Gustry** [comment]
@timlinux Do we need to update the keywords immediately after the download ?
The update is done when these layers has been added to the map canvas (I guess with the signal/slot in the dock).

**Comment by timlinux** [comment]
@gubuntu actually we should be bundling valid keywords in with the zip. I meant we need to update in time with releaseing the next version of InaSAFE. Maybe we can rather add a version get parameter to the flask app so that it bundles a correct version of the keywords based on the version of inasafe running....

**Comment by timlinux** [comment]
Sorry above comment was meant for @Gustry

**Comment by Gustry** [comment]
@timlinux I have a PR on inasafe to provide the version. On the server side, could you check this ticket : https://github.com/kartoza/osm-reporter/issues/62

**Comment by timlinux** [comment]
@Jannes123 LGTM - please merge when ready

File: bin/inasafe [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+usage_file = file(os.path.join(usage_dir, 'usage.txt'))
+for delta in usage_file:
+    usage += delta
+LOGGER = logging.getLogger('InaSAFE')
+QGIS_APP, CANVAS, IFACE, PARENT = get_qgis_app()
+
+
+class CommandLineArguments(object):
+    """Instance objects class for shell arguments.
+    """
```
Comment (timlinux) [inline bin/inasafe]: .. versionadded:: 3.2

File: bin/inasafe [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+            print self.exposure_layers
+        else:
+            self.download = False
+            self.exposure_layers = None
+            LOGGER.debug('no download specified')
+
+
+def download_exposure(command_line_arguments):
+    """Download OSM resources.
+
```
Comment (timlinux) [inline bin/inasafe]: .. versionadded:: 3.2

File: bin/inasafe [hunk 8]
```diff
@@ -0,0 +1,393 @@
...
+        command_line_arguments.exposure)
+    if os.path.exists(command_line_arguments.exposure + '.shp'):
+        print "download successful"
+        command_line_arguments.exposure += '.shp'
+    dialog.close()
+
+
+def get_impact_function_list():
+    """Returns all available impact function ids.
+
```
Comment (timlinux) [inline bin/inasafe]: .. versionadded:: 3.2

**Comment by timlinux** [comment]
@Jannes123 One request is that you annotate all new classes and functions with 

```.. versionadded:: 3.2```

Thanks!

**Pull request merged**